lib/puppet/util/checksums.rb

```diff
@@ -198,6 +195,29 @@ def none(content)
     ""
   end
 
+  class DigestLite
+    def initialize(digest, lite = false)
+      @digest = digest
+      @lite = lite
+      @bytes = 0
+    end
+
+    # Provide an interface for digests. If lite, only digest the first 512 bytes
+    def <<(str)
+      if @lite
+        if (str.length + @bytes) < 512
+          @digest << str
+          @bytes += str.length
+        elsif @bytes < 512
+          @digest << str[0, 512-@bytes]
+          @bytes = 512
+        end
```

Comment:
I think my code and yours are roughly equivalent (if str.length + @bytes == 512 and @bytes < 512 in my case, it falls through to the elsif and does an unnecessary substring (which should be a constant-time operation).

I like your last option though, that should work as well and be simpler to understand.